amanzainal iP

amanzainal opened the pull request:
# Duke

> “Your mind is for having ideas, not holding them.” – David Allen

Duke frees your mind of having to remember things you need to do. It's:

- **Text-based**
- **Easy to learn**
- **FAST SUPER FAST to use**

All you need to do is:

1. Download it from [here](https://not-a-valid-link.com).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

And it is **FREE**!

## Features

- Managing tasks
- Managing deadlines (coming soon)
- Reminders (coming soon)

amanzainal committed:
ignored class files

amanzainal committed:
simple rename, greet and exit

amanzainal committed:
added simple echo logic

amanzainal committed:
fixed text ui

amanzainal committed:
more ui changes

amanzainal committed:
added simple array to store user input and return it

amanzainal committed:
added simple mark task logic

amanzainal committed:
added unmark, bug fixes

amanzainal committed:
added ToDo, Deadline, Event

amanzainal committed:
added automated text UI testing

amanzainal committed:
added error handling

amanzainal committed:
added task deletion

amanzainal committed:
added constants / enums

amanzainal committed:
UI improvements

amanzainal committed:
Add Gradle support

amanzainal committed:
Add Gradle Checkstyle and config files.

Pluiexo commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,22 @@
+
+import duke.exception.*;
+import duke.task.*;
+import duke.ui.*;
```
Try to avoid import all statements as per coding standard guidelines, if this is done by your IDE auto formatter, may I suggest reconfiguring it [in your settings](https://se-education.org/guides/tutorials/intellijCodeStyle.html)

Pluiexo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,103 @@
+                }
+            } else if (arguments.contains("/from") || arguments.contains("/to")) {
+                throw new DukeException("Error: Both /from and /to are required for the event command. Correct format: /from <start> /to <end>");
+            }
```
Interesting way to handle and parse the command string! It may be a bit unintuitive to grasp at first glance, but I suggest a better simplification could be achieved using Pattern Matching with Regular Expressions! You can checkout the Parser class in addressbook2 for reference.

Pluiexo commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,7 @@
+public enum TaskType {
+    T,
+    D,
+    E
```
Wouldn't it be better to give the full name for this task type enumerations?

Pluiexo commented on src/main/java/duke/ui/Messages.java:
```diff
@@ -0,0 +1,39 @@
+            "            |___/                                                      \n" +
+            "\nWhat are we doing today? ~(^o.o^)\nSay 'help' if you want to know the tricks I can do!";
+
+    public static final String EXIT =
```
Awesome Ascii Art!

Pluiexo commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+                // Parse user input
+                Command<List<Task>> parsedCommand = Parser.parseInput(userInput);
+
+                // Execute the command
+                if (parsedCommand != null) {
+                    currentCommand = parsedCommand;
+                    currentCommand.execute(tasks);
+                }
```
Would the null check here be necessary? Rather than nesting your DukeException try catch in Parser, you can bring it up to this level and share this level of try catch, wouldn't that be better? The currentCommand object seems to be a redundant duplicate of parsedCommand too :D

Pluiexo reviewed (commented):
Overall great formatting and naming of classes and variables! Organisations of packages and classes are in order too!

dedsecrattle commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+package duke;
+
+import duke.command.*;
```
Can we avoid using wildcard imports to adhere to the coding style?

dedsecrattle commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,103 @@
+            }
+
+            switch (command) {
+                case "help":
```
No indentation for `case`  clauses in the `switch` , Can we fix this to adhere with the coding standard?

dedsecrattle commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+
+
+public class Duke {
+    public static void main(String[] args) {
```
Can we take out the logic for Running the bot into a separate method e.g  `run` and keep the `main` function as an entry point for the application?

dedsecrattle reviewed (commented):
I like the Abstraction of Commands into separate classes!

HongRay commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+
+import duke.command.*;
+import duke.exception.*;
+import duke.parser.*;
```
Try avoiding wildcard imports as per coding standards!

HongRay commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,22 @@
+import duke.ui.*;
+import java.util.List;
+
+public class AddCommand extends Command<List<Task>> {
```
Add javadocs to your classes and methods other than those trivial methods. These can make reviewing your code easier!

HongRay commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,22 @@
+    public List<Task> execute(List<Task> tasks) {
+        tasks.add(task);
+        Integer count = tasks.size();
+        System.out.printf("\n        ~~~ >^o_o^< ~~~\nGot it! I've added this task:\n - %s\n\nYou have %d task(s) in the list.\n", task, count);
```
Try making your lines shorter, as per coding standards. However, I  like how you tried to make your chat bot cute! 😄

HongRay commented on src/main/java/duke/command/HelpCommand.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public List<Task> execute(List<Task> tasks) {
+        System.out.printf("\nHere's all the things I can do for you! ~(^o.o^)\n");
```
I like your introductory statements. It makes the user knows how to use your chat bot!

HongRay commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,103 @@
+            String start = "Error: No start time set.";
+            String end = "Error: No end time set.";
+
+            // Get Deadline
```
Please give more detailed comments for better clarity as to what you are trying to do in this instance!

HongRay commented on src/main/java/duke/ui/Messages.java:
```diff
@@ -0,0 +1,39 @@
+     */
+    public static final String WELCOME =
+            "  _____            _     _   _             _____      _   _           _   \n" +
+            " / ____|          | |   | | | |           / ____|    | | | |         | |  \n" +
```
Amazing artwork here! Hopefully you can implement this in the GUI!

HongRay reviewed (commented):
Looks good 👍!
Could improve by writing javadocs and some other coding standards.
Great work on the ASCII art! 😄

amanzainal committed:
Remove star imports, reformat and refactor code

amanzainal committed:
Add local data storage

amanzainal committed:
Add LocalDateTime support

amanzainal committed:
Add ViewByDate command

amanzainal committed:
Add JavaDoc comments

amanzainal committed:
Add JavaDoc comments

amanzainal committed:
Format code according to AB3 Checkstyle

amanzainal committed:
Add Find command

amanzainal committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

amanzainal committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

amanzainal committed:
Merge branch 'branch-Level-9'

amanzainal committed:
Format code

amanzainal committed:
Ignore local data files

amanzainal committed:
Ignore local text files

amanzainal committed:
Delete .bugs.txt

Hwww23 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,185 @@
+                }
+            default:
+                throw new DukeException("Sorry.. I don't know what that means!"
+                    + " If you want to leave, just say 'bye'! :(\n");
```
I liked that your switch case adhered to the coding standards! This is one of the most common mistakes made.

Hwww23 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,104 @@
+import java.time.format.DateTimeFormatter;
+import java.util.Scanner;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.ToDo;
```
I like that your import packages are well-separated! It follows the convention of using lowercase letters and separating words with dots.

Hwww23 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    public void setTasks(ArrayList<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    public List<Task> getTasksByDate(LocalDate date) {
+        List<Task> tasksByDate = new ArrayList<>();
+        for (Task task : this.tasks) {
+            if (task instanceof Deadline) {
+                Deadline deadlineTask = (Deadline) task;
+                if (deadlineTask.getBy().toLocalDate().equals(date)) {
+                    tasksByDate.add(task);
+                }
+            } else if (task instanceof Event) {
+                Event eventTask = (Event) task;
+                if (eventTask.getFrom().toLocalDate().equals(date) || eventTask.getTo().toLocalDate().equals(date)) {
+                    tasksByDate.add(task);
+                }
+            }
+        }
+        return tasksByDate;
+    }
+
+
+    /**
+     * Finds tasks in the task list whose description contains a specified keyword.
+     * @param keyword The keyword to search for in task descriptions.
+     * @return A TaskList containing tasks whose descriptions contain the specified keyword.
+     */
+    public TaskList findTasksByKeyword(String keyword) {
+        TaskList matchingTasks = new TaskList();
+        String lowerCaseKeyword = keyword.toLowerCase(); // Convert keyword to lowercase for case insensitivity
+        for (Task task : tasks) {
+            if (task.getDescription().toLowerCase().contains(lowerCaseKeyword)) {
+                matchingTasks.addTask(task);
+            }
+        }
+        return matchingTasks;
+    }
```
I liked that your method names are clear and easy to follow.

Hwww23 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,104 @@
+            } else if (taskType.equals("[D]")) {
+                String dateString = args[3];
+                LocalDateTime dateTime = LocalDateTime.parse(dateString,
+                    DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
```
Perhaps you could consider making the date format a constant to allow for easier changes and reduce repetition.

Hwww23 reviewed (commented):
Overall, it looks well done!

amanzainal committed:
Add Dummy Test file

amanzainal committed:
Add test class for Mark Command

amanzainal committed:
Restructure code to remove "arrowhead" style

amanzainal committed:
Add catch block

amanzainal committed:
Add JUnit testing for Mark Command

amanzainal committed:
Remove unused import

amanzainal committed:
Remove unused import

amanzainal committed:
Add JUnit testing for Delete Command

amanzainal committed:
Create JAR file

amanzainal committed:
Add flexible constructors

amanzainal committed:
Add method to get all tasks

amanzainal committed:
Fix storage not handling missing data files

amanzainal committed:
Fix file not found issue

amanzainal committed:
Add varargs

amanzainal committed:
Merge branch 'master' of https://github.com/amanzainal/ip

amanzainal committed:
Add JavaFX support

amanzainal committed:
Add Launcher class

amanzainal committed:
Add layout design

amanzainal committed:
Add control styling

amanzainal committed:
Add echo functionality and autoscroll

amanzainal committed:
Fix autoscroll feature

amanzainal committed:
Add dialog boxes

amanzainal committed:
Add flipped-facing dialog boxes

amanzainal committed:
Add flipped-facing dialog boxes

amanzainal committed:
Add JavaDoc comments

amanzainal committed:
Add padding between DialogBoxes

amanzainal committed:
Add padding between ImageView and Label

amanzainal committed:
Add circle clipping for ImageView

amanzainal committed:
Add background color to DialogBox

amanzainal committed:
Add FXML controllers

amanzainal committed:
Restructure file structure

amanzainal committed:
Fix GUI

amanzainal committed:
Refactor print statements

amanzainal committed:
Fix minor formatting issues

amanzainal committed:
Refactor codebase

amanzainal committed:
Merge branch 'branch-Level-10'

amanzainal committed:
Add assertions for improved code reliability

- Ensure task is not null before adding it to the task list
- Assert that index for accessing tasks is within bounds
- Add assertion to ensure response message is not null in UI
- Assert that message is not null before appending it to UI response

These assertions enhance code reliability by checking critical
assumptions at various points in the codebase, reducing the risk of
unexpected behavior.

amanzainal committed:
Refactor Parser to Improve Readability and Maintainability

This commit refactors the `Parser.java` file to enhance code readability
and maintainability by applying the Single Level of Abstraction Principle (SLAP).

Changes Made:
- Extracted deadline parsing logic into a separate method called `parseDeadline`.
  This method handles the parsing of deadline descriptions and date/time information.
- Extracted event parsing logic into a separate method called `parseEvent`.
  This method handles the parsing of event descriptions and start/end date/time information.
- Simplified the switch case in the `parseInput` method by using the extracted methods for adding tasks.
  This results in a cleaner and more concise switch case structure.

These changes significantly reduce the complexity of the `parseInput` method while maintaining clarity and readability.

amanzainal committed:
Refactor HelpCommand to use streams for message formatting

This commit updates the HelpCommand class to utilize streams for formatting the list of available commands. Instead of manually appending newline characters between each command, the messages are joined using the Collectors.joining() method, resulting in a cleaner and more concise implementation. This improves the readability and maintainability of the codebase.

Changes Made:
- Refactored the HelpCommand class to use streams for message formatting
- Updated the execute() method to utilize streams for concatenating help messages

amanzainal committed:
Update TaskList to utilize streams for task manipulation

This commit enhances the TaskList class by leveraging Java streams for various task-related operations, improving code readability and maintainability.

Changes Made:
- Refactored getTasksAsString() method to use streams for building the string representation of tasks.
- Updated getTasksByDate() method to utilize streams for filtering tasks by date.
- Improved findTasksByKeyword() method by employing streams for searching tasks by keyword.

amanzainal committed:
Merge pull request #2 from amanzainal/branch-A-Streams

Merge A-Streams branch into master

amanzainal committed:
Merge pull request #4 from amanzainal/branch-A-Assertions

Merge A-Assertions branch into master

amanzainal committed:
Merge pull request #3 from amanzainal/branch-A-CodeQuality

Merge A-CodeQuality branch into master

amanzainal committed:
Add friendlier syntax

amanzainal committed:
Improve GUI readability, fix long text messages not displaying properly

amanzainal committed:
Merge branch 'branch-A-BetterGUI'

amanzainal committed:
Add new user / bot images

amanzainal committed:
Abstract away user-facing strings

amanzainal committed:
Temporary state-save commit

amanzainal committed:
Change date format for ViewByDate command

amanzainal committed:
Minor bug fix

amanzainal committed:
Add send icon

amanzainal committed:
Add stage icon

amanzainal committed:
Remove hidden newline characters

amanzainal committed:
Abstract away user-facing strings

amanzainal committed:
Restructure code to improve readability

amanzainal committed:
Merge remote-tracking branch 'origin/branch-A-Personality'

amanzainal committed:
Merge branch 'branch-A-Personality'

amanzainal committed:
Fix test classes

amanzainal committed:
Fix backwards compatibility issues

amanzainal committed:
Add JavaDoc comments to pass Checkstyle

amanzainal committed:
Add more command aliases

amanzainal committed:
Add Clear command to delete all tasks

amanzainal committed:
Add Noop (No-operation) command to do nothing

amanzainal committed:
Add exit command to close application

amanzainal committed:
Fix minor formatting issues

amanzainal committed:
Fix error messages not displaying properly

amanzainal committed:
Fix potential bug with event parsing

amanzainal committed:
Fix potential bug with ViewByDate parsing

amanzainal committed:
Update user-facing strings

amanzainal committed:
Add utility method to quit application

amanzainal committed:
Remove redundant comment

amanzainal committed:
Test application exit

amanzainal committed:
Revert "Test application exit"

This reverts commit e5ab3205f3a034983acb9abe838ecb8ae9e7f9d7.

amanzainal committed:
Replace all "Duke" to "Coat"

amanzainal committed:
Add product showcase

amanzainal committed:
Update README.md

amanzainal committed:
Update README.md

amanzainal committed:
Update README.md

amanzainal committed:
Add quick fix for empty todo description

amanzainal committed:
Merge branch 'branch-fix-todo'

amanzainal committed:
Remove I/O redirection test support

amanzainal committed:
Add Continuous Integration (CI) support

amanzainal committed:
Remove debug print statements

amanzainal committed:
Remove remote tasks file

amanzainal committed:
Remove I/O redirection test support

amanzainal committed:
Change gradlew to executable to fix CI flow
